src/ring_theory/hahn_series.lean

```diff
@@ -0,0 +1,390 @@
+  * If `R` is a (comm_)(semi)ring, then so is `hahn_series Γ R`.
+
+## TODO
+  * Given `[linear_ordered_add_comm_group Γ]` and `[field R]`, define `field (hahn_series Γ R)`.
+  * Build an API for the constant map `C` and the variable `X`
+  * Define Laurent series
+
+-/
+
+open_locale big_operators
+
+/-- If `Γ` is linearly ordered and `R` has zero, then `hahn_series Γ R` consists of
+  formal series over `Γ` with coefficients in `R`, whose supports are well-founded. -/
+@[ext]
+structure hahn_series (Γ : Type*) (R : Type*) [linear_order Γ] [has_zero R] :=
+(coeff : Γ → R)
+(is_wf_support : {a | coeff a ≠ 0}.is_wf)
+
+variables {Γ : Type*} {R : Type*}
+
+namespace hahn_series
+
+section zero
+variables [linear_order Γ] [has_zero R]
+
+instance : has_zero (hahn_series Γ R) :=
+⟨{ coeff := 0,
+   is_wf_support := by simp }⟩
+
+instance : inhabited (hahn_series Γ R) := ⟨0⟩
+
+@[simp]
+lemma zero_coeff {a : Γ} : (0 : hahn_series Γ R).coeff a = 0 := rfl
+
+/-- `single a r` is the Hahn series which has coefficient `r` at `a` and zero otherwise. -/
+def single (a : Γ) (r : R) : hahn_series Γ R := {
+  coeff := λ b, if a = b then r else 0,
+  is_wf_support := (set.is_wf_singleton a).mono (λ b, begin
+    simp only [and_imp, exists_prop, set.mem_singleton_iff, ite_eq_right_iff,
+      ne.def, set.mem_set_of_eq, not_forall],
+    exact λ h1 _, h1.symm,
+  end) }
+
+variables {a b : Γ} {r : R}
+
+theorem single_coeff : (single a r).coeff b = if a = b then r else 0 := rfl
+
+@[simp]
+theorem single_coeff_same : (single a r).coeff a = r := if_pos rfl
+
+@[simp]
+theorem single_coeff_of_ne (h : a ≠ b) : (single a r).coeff b = 0 := if_neg h
+
+@[simp]
+lemma single_support_of_ne (h : r ≠ 0) : {b | (single a r).coeff b ≠ 0} = {a} :=
+begin
+  ext b,
+  simp only [set.mem_singleton_iff, ne.def, set.mem_set_of_eq],
+  split,
+  { contrapose!,
+    exact λ h', single_coeff_of_ne (ne.symm h') },
+  { rintro rfl,
+    simp [h] }
+end
+
+@[simp]
+lemma single_eq_zero : (single a (0 : R)) = 0 := by { ext, simp [single_coeff] }
+
+end zero
+section addition
+
+variable [linear_order Γ]
+
+instance [add_monoid R] : has_add (hahn_series Γ R) :=
+{ add := λ x y, { coeff := x.coeff + y.coeff,
+                  is_wf_support := (x.is_wf_support.union y.is_wf_support).mono (λ a ha, begin
+                    contrapose! ha,
+                    simp only [not_or_distrib, not_not, set.mem_union_eq, set.mem_set_of_eq] at ha,
+                    simp only [pi.add_apply, not_not, set.mem_set_of_eq],
+                    rw [ha.1, ha.2, add_zero],
+                  end) } }
+
+instance [add_monoid R] : add_monoid (hahn_series Γ R) :=
+{ zero := 0,
+  add := (+),
+  add_assoc := λ x y z, by { ext, apply add_assoc },
+  zero_add := λ x, by { ext, apply zero_add },
+  add_zero := λ x, by { ext, apply add_zero } }
+
+@[simp]
+lemma add_coeff [add_monoid R] {x y : hahn_series Γ R} {a : Γ} :
+  (x + y).coeff a = x.coeff a + y.coeff a := rfl
```

Comment:
Perhaps
```suggestion
@[simp]
lemma add_coeff [add_monoid R] {x y : hahn_series Γ R} {a : Γ} :
  (x + y).coeff = x.coeff + y.coeff := rfl
```
or both,
```suggestion
@[simp]
lemma add_coeff [add_monoid R] {x y : hahn_series Γ R} {a : Γ} :
  (x + y).coeff = x.coeff + y.coeff := rfl

lemma add_coeff_apply [add_monoid R] {x y : hahn_series Γ R} {a : Γ} :
  (x + y).coeff a = x.coeff a + y.coeff a := rfl
```

Reply:
I've gone with `add_coeff'` and `add_coeff` instead of `add_coeff` and `add_coeff_apply`, and similar for neg and sub.